File: app/src/main/java/fr/socket/florian/dhome/view/about/AboutAdapter.java
package fr.socket.florian.dhome.view.about;

import android.support.annotation.NonNull;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import java.util.ArrayList;
import java.util.List;

import fr.socket.florian.dhome.R;

class AboutAdapter extends RecyclerView.Adapter<RecyclerView.ViewHolder> {

    private static final int ICON_VIEW_TYPE = 1;
    private static final int HEADER_VIEW_TYPE = 2;
    private static final int GITHUB_VIEW_TYPE = 3;

    private final List<Integer> headers;
    private final List<Icon> icons;
    private final List<Github> githubs;

    private int headerCounter;
    private int iconsCounter;
    private int githubsCounter;

    AboutAdapter() {
        icons = new ArrayList<>();
        icons.add(new Icon(R.drawable.ic_light_bulb, "Icon made by Freepik from www.flaticon.com ", "http://www.freepik.com/"));
        icons.add(new Icon(R.drawable.ic_server, "Icon made by Smashicons from www.flaticon.com ", "https://smashicons.com/"));
        icons.add(new Icon(R.drawable.ic_user, "Icon made by Smashicons from www.flaticon.com ", "https://smashicons.com/"));
        icons.add(new Icon(R.drawable.ic_locked, "Icon made by Smashicons from www.flaticon.com ", "https://smashicons.com/"));
        headers = new ArrayList<>();
        headers.add(R.string.repositories);
        headers.add(R.string.icons);
        githubs = new ArrayList<>();
        githubs.add(new Github("Restful-home", "Node js backend API to control the home devices through a raspberry pi", "https://github.com/FlorianArnould/restful-home"));
        githubs.add(new Github("D-Home", "An android application to use the restful-home backend API", "https://github.com/FlorianArnould/D-Home"));
    }

    @Override
    public int getItemViewType(int position) {
        if (position == 0 || position == githubs.size() + 1) {
            return HEADER_VIEW_TYPE;
        } else if (position <= githubs.size()) {
            return GITHUB_VIEW_TYPE;
        }
        return ICON_VIEW_TYPE;
    }

    @NonNull
    @Override
    public RecyclerView.ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View view;
        switch (viewType) {
            case HEADER_VIEW_TYPE:
                view = LayoutInflater.from(parent.getContext()).inflate(R.layout.header_row, parent, false);
                return new HeaderViewHolder(view);
            case ICON_VIEW_TYPE:
                view = LayoutInflater.from(parent.getContext()).inflate(R.layout.icon_row, parent, false);
                return new IconViewHolder(view);
            case GITHUB_VIEW_TYPE:
                view = LayoutInflater.from(parent.getContext()).inflate(R.layout.github_row, parent, false);
                return new GithubViewHolder(view);
        }
        view = LayoutInflater.from(parent.getContext()).inflate(R.layout.icon_row, parent, false);
        return new RecyclerView.ViewHolder(view) {
            @Override
            public String toString() {
                return super.toString();
            }
        };
    }

    @Override
    public void onBindViewHolder(@NonNull RecyclerView.ViewHolder viewHolder, int position) {
        if (position == 0) {
            headerCounter = 0;
            iconsCounter = 0;
            githubsCounter = 0;
        }
        switch (getItemViewType(position)) {
            case ICON_VIEW_TYPE:
                ((IconViewHolder) viewHolder).update(icons.get(iconsCounter++));
                break;
            case HEADER_VIEW_TYPE:
                ((HeaderViewHolder) viewHolder).update(headers.get(headerCounter++));
                break;
            case GITHUB_VIEW_TYPE:
                ((GithubViewHolder) viewHolder).update(githubs.get(githubsCounter++));
                break;
        }
    }

    @Override
    public int getItemCount() {
        return icons.size() + headers.size() + githubs.size();
    }
}

File: app/src/main/java/fr/socket/florian/dhome/view/about/AboutAdapter.kt
package fr.socket.florian.dhome.view.about

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.recyclerview.widget.RecyclerView

import fr.socket.florian.dhome.R

internal class AboutAdapter : RecyclerView.Adapter<RecyclerView.ViewHolder>() {

    private val headers: List<Int> = listOf(R.string.repositories, R.string.icons)
    private val icons: List<Icon> = listOf(
            Icon(R.drawable.ic_light_bulb, "Icon made by Freepik from www.flaticon.com ", "http://www.freepik.com/"),
            Icon(R.drawable.ic_server, SMASHICONS_TEXT, SMASHICONS_URL),
            Icon(R.drawable.ic_user, SMASHICONS_TEXT, SMASHICONS_URL),
            Icon(R.drawable.ic_locked, SMASHICONS_TEXT, SMASHICONS_URL))
    private val githubs: List<Repository> = listOf(
            Repository("Restful-home", "Node js backend API to control the home devices through a raspberry pi", "https://github.com/FlorianArnould/restful-home"),
            Repository("D-Home", "An android application to use the restful-home backend API", "https://github.com/FlorianArnould/D-Home"))

    private var headerCounter: Int = 0
    private var iconsCounter: Int = 0
    private var repositoriesCounter: Int = 0

    override fun getItemViewType(position: Int): Int {
        if (position == 0 || position == githubs.size + 1) {
            return HEADER_VIEW_TYPE
        } else if (position <= githubs.size) {
            return GITHUB_VIEW_TYPE
        }
        return ICON_VIEW_TYPE
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder {
        val view: View
        when (viewType) {
            HEADER_VIEW_TYPE -> {
                view = LayoutInflater.from(parent.context).inflate(R.layout.header_row, parent, false)
                return HeaderViewHolder(view)
            }
            ICON_VIEW_TYPE -> {
                view = LayoutInflater.from(parent.context).inflate(R.layout.icon_row, parent, false)
                return IconViewHolder(view)
            }
            GITHUB_VIEW_TYPE -> {
                view = LayoutInflater.from(parent.context).inflate(R.layout.repository_row, parent, false)
                return RepositoryViewHolder(view)
            }
        }
        view = LayoutInflater.from(parent.context).inflate(R.layout.icon_row, parent, false)
        return object : RecyclerView.ViewHolder(view) {}
    }

    override fun onBindViewHolder(viewHolder: RecyclerView.ViewHolder, position: Int) {
        if (position == 0) {
            headerCounter = 0
            iconsCounter = 0
            repositoriesCounter = 0
        }
        when (getItemViewType(position)) {
            ICON_VIEW_TYPE -> (viewHolder as IconViewHolder).update(icons[iconsCounter++])
            HEADER_VIEW_TYPE -> (viewHolder as HeaderViewHolder).update(headers[headerCounter++])
            GITHUB_VIEW_TYPE -> (viewHolder as RepositoryViewHolder).update(githubs[repositoriesCounter++])
        }
    }

    override fun getItemCount(): Int {
        return icons.size + headers.size + githubs.size
    }

    companion object {
        private const val SMASHICONS_TEXT = "Icon made by Smashicons from www.flaticon.com "
        private const val SMASHICONS_URL = "https://smashicons.com/"
        private const val ICON_VIEW_TYPE = 1
        private const val HEADER_VIEW_TYPE = 2
        private const val GITHUB_VIEW_TYPE = 3
    }
}
